firedrake/preconditioners/fdm.py

```diff
@@ -0,0 +1,1036 @@
+
+        if len(self.bcs) > 0:
+            self.bc_nodes = numpy.unique(numpy.concatenate([bcdofs(bc, ghost=False)
+                                                            for bc in self.bcs]))
+        else:
+            self.bc_nodes = numpy.empty(0, dtype=PETSc.IntType)
+
+        bcflags = self.get_bc_flags(V, self.mesh, self.bcs, solverctx._problem.J)
+
+        self.weight = self.multiplicity(V)
+        with self.weight.dat.vec as w:
+            w.reciprocal()
+
+        # Get form coefficients
+        appctx = self.get_appctx(pc)
+        eta = appctx.get("eta", (N+1)*(N+ndim))  # interior penalty parameter
+        mu = appctx.get("viscosity", None)  # second order coefficient
+        beta = appctx.get("reaction", None)  # zeroth order coefficient
+        eta = float(eta)
+        self.appctx = appctx
+
+        # Obtain the FDM basis and transfer kernels (restriction and prolongation)
+        # Afdm = 1D interval stiffness and mass matrices in the FDM basis for each direction and BC type
+        # Dfdm = normal derivate matrices in the FDM basis
+        Afdm, Dfdm, self.restrict_kernel, self.prolong_kernel = self.assemble_matfree(V, N, Nq, eta, needs_interior_facet)
+
+        # Preallocate by calling the assembly routine on a PETSc Mat of type PREALLOCATOR
+        prealloc = PETSc.Mat().create(comm=A.comm)
+        prealloc.setType(PETSc.Mat.Type.PREALLOCATOR)
+        prealloc.setSizes(A.getSizes())
+        prealloc.setUp()
+        ndof = V.value_size * V.dof_dset.set.size
+
+        # PDE coefficients interpolated on the quadrature nodes
+        ele = V.ufl_element()
+        ncomp = ele.value_size()
+        bsize = V.value_size
+        needs_hdiv = bsize != ncomp
+        Gq, Bq, self._assemble_Gq, self._assemble_Bq = self.assemble_coef(mu, beta, Nq, diagonal=True, piola=needs_hdiv)
+
+        # Assign arbitrary non-zero coefficients for preallocation
+        Gq.dat.data[:] = 1.0E0
+        if Bq is not None:
+            Bq.dat.data[:] = 1.0E0
+
+        self.assemble_kron(prealloc, V, Gq, Bq, Afdm, Dfdm, eta, bcflags, needs_interior_facet)
+        nnz = get_preallocation(prealloc, ndof)
+        self.Pmat = PETSc.Mat().createAIJ(A.getSizes(), nnz=nnz, comm=A.comm)
+        self._assemble_Pmat = partial(self.assemble_kron, self.Pmat, V, Gq, Bq,
+                                      Afdm, Dfdm, eta, bcflags, needs_interior_facet)
+
+        prealloc.destroy()
+        lgmap = V.dof_dset.lgmap
+        self.Pmat.setBlockSize(V.value_size)
+        self.Pmat.setLGMap(lgmap, lgmap)
+
+        opc = pc
+        # Internally, we just set up a PC object that the user can configure
+        # however from the PETSc command line.  Since PC allows the user to specify
+        # a KSP, we can do iterative by -fdm_pc_type ksp.
+        pc = PETSc.PC().create(comm=opc.comm)
+        pc.incrementTabLevel(1, parent=opc)
+
+        # We set a DM on the constructed PC so one
+        # can do patch solves with ASMPC.
```

Comment:
Do you need a DM for `PCASM`? I thought it was purely algebraic.

Reply:
Yes, it needs the DM to define ISes